Next release features

VladimirMakeev opened the pull request:
Battle action script replaces vanilla logic with a custom one. Script allows to choose action of a unit controlled by AI either it is a non-human player, Neutral race or 'fast battle' mode.
Scripts are assigned to AI attitude and specified in a new column `ACTION_S` in `GAI.dbf`.
While player AI attitudes can be specified in Scenario Editor capital menu, neutral player attitude now can be specified in scenario properties menu.

VladimirMakeev committed:
Add lua bindings for GVars.dbf

VladimirMakeev committed:
Add lua bindings for GlobalData

VladimirMakeev committed:
Fix typo

VladimirMakeev committed:
Add lua bindings for game rules and user settings

VladimirMakeev committed:
Support id types in lua api

VladimirMakeev committed:
Update group lua api

VladimirMakeev committed:
Update stack and unitimpl lua api

VladimirMakeev committed:
Add missing include

VladimirMakeev committed:
Fix const

VladimirMakeev committed:
Update lua bindings for scenario

VladimirMakeev committed:
Add CMidCrystal structure

VladimirMakeev committed:
Add lua bindings for crystals

VladimirMakeev committed:
Return capital protection when calling prot(6)

VladimirMakeev committed:
Add CMidSiteMerchant structure

VladimirMakeev committed:
Add CMidSiteMage structure

VladimirMakeev committed:
Add CMidSiteMercs structure

VladimirMakeev committed:
Add CMidSiteTrainer structure

VladimirMakeev committed:
Add lua bindings for sites

VladimirMakeev committed:
Add lua bindings for merchant

VladimirMakeev committed:
Add lua bindings for mercenary camps

VladimirMakeev committed:
Add lua bindings for trainer

VladimirMakeev committed:
Allow access to sites from scenario in lua api

VladimirMakeev committed:
Place ai attitude categories separately

VladimirMakeev committed:
Add CAiAttitudesTable structure and its api

VladimirMakeev committed:
Add visitor to change player attitude

VladimirMakeev committed:
Add CScenPropInterf structure and its api

VladimirMakeev committed:
Allow changing Netrals attitude from scenario properties menu

VladimirMakeev committed:
Add function to get unit group owner

VladimirMakeev committed:
Add new functions to BattleMsgData api

VladimirMakeev committed:
Add lua bindings for mutable BattleMsgData

VladimirMakeev committed:
Load custom battle action scripts for AI

VladimirMakeev committed:
Support custom scripts for AI battle actions

VladimirMakeev committed:
Update lua api docs

VladimirMakeev committed:
Fix typos

VladimirMakeev committed:
Add entrance method to fort binding

VladimirMakeev committed:
Add AI battle action script example

VladimirMakeev committed:
Add fast battle to battle api

VladimirMakeev committed:
Support battle turns order in lua api

VladimirMakeev committed:
Update battle api

VladimirMakeev committed:
Update scenario api

VladimirMakeev committed:
Add method to check unit wait status in battle

VladimirMakeev committed:
Check attack existence before accessing it

VladimirMakeev committed:
Event system optimization

VladimirMakeev committed:
'Kill stack' event condition optimization

VladimirMakeev committed:
'Own item' event condition optimization

VladimirMakeev committed:
Describe CMidgardPlan implementation

VladimirMakeev committed:
Implement cache for stack templates

Cache is used to track which stacks were created from which stack
template in order to speed up event conditions that depend on them.
Cache is intended to work only on server side, as its purpose is to
speed up the event system.
Cache is cleared and set during scenario loading. Cache is updated each
time stack is removed or source template id is assigned.

VladimirMakeev committed:
'Leader own item' event condition optimization

Instead of searching among all scenario objects for stack created
from template we only search in stack template cache.

VladimirMakeev committed:
'Stack exists' event condition optimization

VladimirMakeev committed:
Add Lua api documentation in LaTeX

VladimirMakeev committed:
Fix battle actions script

VladimirMakeev committed:
Update gitignore for LaTeX

VladimirMakeev committed:
Implement resource market site

Resource market is a new site object for stacks to visit
on a strategic map. Resource market allows player to exchange
gold and mana resources according to exchange rates. Rates
are defined using Lua scripts and can be specific for an object or
use default rules that are global for the whole game.

UnveN commented on mss32/src/bindings/gameview.cpp:
```diff
@@ -0,0 +1,55 @@
+
+int GameView::getUnitMinDamage() const
+{
+    return 1;
```
Better to read from GameRestrictions::attackDamage[0]

VladimirMakeev replied:
Done

UnveN commented on mss32/src/bindings/gameview.cpp:
```diff
@@ -0,0 +1,55 @@
+
+int GameView::getLeaderAdditionalDamage() const
+{
+    return 100;
```
Better to read from GameRestrictions::fighterExplorerLeaderBonusMaxDamage

VladimirMakeev replied:
Done

UnveN commented on mss32/include/bindings/unitimplview.h:
```diff
@@ -99,8 +99,10 @@ class UnitImplView
+    /** Returns alternative attack for primary attack. */
     std::optional<AttackView> getAltAttack() const;
+    /** Returns alternative attack for secondary attack. */
+    std::optional<AttackView> getAltAttack2() const;
```
Not sure if we really want to expose this in a public API - as this is basically unused in game mechanics.

VladimirMakeev replied:
It is used deep inside [battle logic](https://github.com/VladimirMakeev/D2ModdingToolset/blob/f846931530bfe61cfd64f024a2aa88c66f13dc58/Scripts/battleAi.lua#L1305)

UnveN commented on mss32/src/bindings/scenarioview.cpp:
```diff
@@ -469,6 +509,140 @@ std::optional<DiplomacyView> ScenarioView::getDiplomacy() const
     return DiplomacyView{hooks::getDiplomacy(objectMap)};
 }
 
+void ScenarioView::forEachStack(const std::function<void(const StackView&)>& callback) const
```
This whole family of forEachXXX looks like a perfect candidate for template<> implementation to avoid copy-pasting.

VladimirMakeev replied:
Not so easy because of game specific dynamic casts and different number of arguments for different views. I think this approach is easier to read and reason about each case than generic templates

UnveN commented on mss32/src/bindings/globalvariablesview.cpp:
```diff
@@ -348,10 +348,14 @@ int GlobalVariablesView::getDBkaTurn() const
 int GlobalVariablesView::getProt(int tier) const
 {
-    if (tier < 1 || tier > 5) {
+    if (tier < 1 || tier > 6) {
```
If we want strict matching to game internals, then a capital is not a tier 6 city but a distinct type of fortification. I think it would be better to split its armor value into its own property.

VladimirMakeev replied:
Modmakers asked for a nicer api without if-else around city tiers and various buff levels whenever possible. I also noticed that `CCapital::getTier()` returns 6.

UnveN commented on mss32/src/bindings/siteview.cpp:
```diff
@@ -0,0 +1,61 @@
+{
+    auto view = lua.new_usertype<SiteView>("SiteView");
+
+    bindAccessMethods(view);
```
Maybe making SiteView::bind virtual and then calling it in descendant implementations would be more straightforward solution.

UnveN commented on mss32/src/visitors.cpp:
```diff
@@ -36,6 +36,7 @@ std::array<Api, 3> functions = {{
         (Api::TransformUnit)0x5e968e,
         (Api::UndoTransformUnit)0x5e96df,
         (Api::ExtractUnitFromGroup)0x5e8d72,
+        (Api::PlayerSetAttitude)0,
```
nullptr looks better

VladimirMakeev replied:
Done

UnveN commented on mss32/src/bindings/battlemsgdataview.cpp:
```diff
@@ -36,13 +39,7 @@ BattleMsgDataView::BattleMsgDataView(const game::BattleMsgData* battleMsgData,
-    view["defenderPlayer"] = sol::property(&BattleMsgDataView::getDefenderPlayer);
-    view["attacker"] = sol::property(&BattleMsgDataView::getAttacker);
-    view["defender"] = sol::property(&BattleMsgDataView::getDefender);
+    bindAccessMethods(view);
```
Maybe its better to use virtual ::bind of base class.

UnveN commented on mss32/src/bindings/battlemsgdataview.cpp:
```diff
@@ -92,6 +89,286 @@ std::optional<GroupView> BattleMsgDataView::getDefender() const
+        setApi.destructor((IntSet*)&battleActions);
+    }
+
+    return {actions,                          //
```
What are these empty comments?

VladimirMakeev replied:
I used them to change code formatting, removed

UnveN commented on mss32/src/bindings/battlemsgdataview.cpp:
```diff
@@ -92,6 +89,286 @@ std::optional<GroupView> BattleMsgDataView::getDefender() const
+
+bool BattleMsgDataView::isUnitResistantToSourceById(const IdView& unitId, int sourceId) const
+{
+    using namespace game;
```
Better to move this logic into deicated util function to lighten binding code and make the logic accessible for further usage.

VladimirMakeev replied:
Done

UnveN commented on mss32/src/bindings/battlemsgdataview.cpp:
```diff
@@ -92,6 +89,286 @@ std::optional<GroupView> BattleMsgDataView::getDefender() const
+    const auto& attackClasses{AttackClassCategories::get()};
+    const LAttackClass* attackClass = nullptr;
+
+    const AttackClassId id{static_cast<AttackClassId>(classId)};
```
Better to move this code to dedicated util getAttackClassById

VladimirMakeev replied:
Done

UnveN commented on mss32/src/scripts.cpp:
```diff
@@ -283,6 +283,23 @@ static void bindApi(sol::state& lua)
+        "Defend", BattleAction::Defend,
+        "Auto", BattleAction::Auto,
+        "UseItem", BattleAction::UseItem
+        // Restrict access to Resolve action from scripts
```
Better to explicitly deny the Resolve action from desired script functions that to completely remove its declaration from api.

VladimirMakeev replied:
I could not think of a single use case for `Resolve` action in scripts. For fast and auto battle modes we have appropriate properties in `BattleView` and AI does not touch `Resolve` action in its logic.

UnveN replied:
@VladimirMakeev Yeah, I agree that Resolve have no usage in scripts just now. But it looks like a crutch to me to hide parts of exposed game data from API just because it has no usage right now. Just imho, ofcourse )

VladimirMakeev replied:
If we find out that we need this action - it is easy to update api. Easier than introducing it now and telling everyone not to use it.

UnveN commented on Scripts/settings.lua:
```diff
@@ -243,6 +243,10 @@ settings = {
 		carryXpOverUpgrade = false,
 		-- Allows units to receive multiple upgrades per single battle
 		allowMultiUpgrade = false,
+        -- Debug battle AI action script
```
Replace spaces with tabs.

VladimirMakeev replied:
Done

UnveN commented on Scripts/settings.lua:
```diff
@@ -243,6 +243,10 @@ settings = {
 		-- Allows units to receive multiple upgrades per single battle
 		allowMultiUpgrade = false,
+        -- Debug battle AI action script
+        debugAi = true,
```
It is not clear what this flag really does and how it correlates with debugMode enabled vs disabled.

VladimirMakeev replied:
Updated description

UnveN commented on mss32/include/battlemsgdatahooks.h:
```diff
@@ -27,6 +27,7 @@ namespace game {
 struct CMidgardID;
 struct BattleMsgData;
 struct IMidgardObjectMap;
+struct PossibleTargets;
```
Probably not the best name for the data type. Because the same set of targets can be used in different contexts in various functions. For instance "possible targets" in one function can become "invalid targets" in another. More neutral name would be better. Consider names like "AttackTargets" or "UnitAttackTargets". Since all battle item effects are considered as "attacks", simple "AttackTargets" is sufficient.

UnveN commented on mss32/src/bindings/fortview.cpp:
```diff
@@ -57,6 +58,14 @@ Point FortView::getPosition() const
+    const auto& mapElement{fort->mapElement};
+    const auto& position{mapElement.position};
+
+    return Point(position.x + mapElement.sizeX - 1, position.y + mapElement.sizeY - 1);
```
While the function itself is simple, the method of entrance calculation is useful for further usage. Consider moving it to a util function. Another point is that bindings itself should not carry any additional logic other than being a thin wrapper over game objects.

VladimirMakeev replied:
Moved to separate function

UnveN commented on mss32/src/battlemsgdatahooks.cpp:
```diff
@@ -447,23 +447,6 @@ void __stdcall aiChooseBattleActionHooked(const game::IMidgardObjectMap* objectM
     }
 
-    // This should never happen
-    if (allyPlayer->isHuman) {
```
It really happened during auto-battle, right? )

VladimirMakeev replied:
Yes 🤣

UnveN commented on mss32/src/unitutils.cpp:
```diff
@@ -242,6 +242,10 @@ game::IAttack* getAttack(const game::IUsUnit* unit, bool primary, bool checkAltA
 game::IAttack* getAltAttack(const game::IUsUnit* unit, bool primary)
 {
     auto attack = getAttack(unit, primary, false);
+    if (!attack) {
```
While the check is admittedly not excessive, are there any occurrences where a unit can not have a primary attack?

UnveN commented on mss32/src/midserverlogichooks.cpp:
```diff
@@ -146,4 +161,409 @@ bool __fastcall midServerLogicSendRefreshInfoHooked(const game::CMidServerLogic*
+                                                                              playingStackId);
+}
+
+bool __fastcall stackMoveHooked(game::CMidServerLogic** thisptr,
```
Did you really manage to make this shit work faster? (O_O)

UnveN commented on mss32/include/midgardplan.h:
```diff
@@ -39,13 +39,35 @@ assert_size(MidgardPlanElement, 8);
+     * Each array element stores flags of 4 adjacent tiles along X axis.
+     * Accessed as: 36 * Y + X / 4.
+     */
+    std::uint8_t elementFlags[5184];
```
Are you sure its really laying out like that? More likely its 144 x 144 x 2 bits (exactly 5184 bytes). So, for instance, the first bit might be indicating if there is a static element on a tile, while the second - is there a dynamic.

VladimirMakeev replied:
Formula was taken directly from game logic

VladimirMakeev replied:
I decided to double check, here is the example of access:
```c++
// 0x5f6819 (Akella, Russobit)
void __stdcall MidgardPlanGetStaticDynamicElementFlags(
        unsigned __int8 *data,
        signed int positionX,
        int positionY,
        bool *staticElement,
        bool *dynamicElement)
{
  unsigned __int8 value; // cl

  value = data[36 * positionY + (positionX >> 2)];
  *staticElement = 0;
  *dynamicElement = 0;
  
  // unsigned char midgardPlanStaticElementMasks[8] = { 2, 1, 8, 4, 0x20, 0x10, 0x80, 0x40 };
  if ( (value & midgardPlanStaticElementMasks[2 * (positionX & 3)]) != 0 )
    *staticElement = 1;
    
  // unsigned char midgardPlanDynamicElementMasks[7] = { 1, 8, 4, 0x20, 0x10, 0x80, 0x40 };
  if ( (value & midgardPlanDynamicElementMasks[2 * (positionX & 3)]) != 0 )
    *dynamicElement = 1;
}
```
As you can see, bitmask is applied to whole `unsigned char` value and mask choice depends on `positionX`.
Of course, the compiler could have optimized the logic in such way, we'll never know.
I don't think we would ever touch these values by hand, only using original game functions.
I left a comment so that the general idea becomes cleaner and the knowledge not lost.

UnveN replied:
Yeah, if we ever want to find out, we could just inspect the bits on debugging to see how it correlates with a current map layout.

UnveN commented on mss32/src/hooks.cpp:
```diff
@@ -2433,4 +2449,33 @@ game::BuildingStatus __stdcall getBuildingStatusHooked(const game::IMidgardObjec
     return BuildingStatus::CanBeBuilt;
 }
 
+bool __stdcall removeStackHooked(const game::CMidgardID* stackId,
```
Might be an excessive operation in terms of performance since stack id counter is incremented each time a new stack is created - so there should not be any id collision during a single scenario, while access time should be amortized O(1).
In the other hand, however, is that the "remove stack" event should normally happen rarely comparing to event condition check callbacks. So nevermind ;)

VladimirMakeev committed:
Update resource market

- support resource market in scenario generator
- support market in Lua api
- add scenario difficulty to Lua api
- update generator and Lua api documentation

VladimirMakeev committed:
Add missing hooks

VladimirMakeev committed:
Define CMidClientData fields

VladimirMakeev committed:
Add location name to lua api

VladimirMakeev committed:
Add editor check to Lua api

VladimirMakeev committed:
Code review fixes

VladimirMakeev committed:
Move entrance computation to utility function

VladimirMakeev committed:
Code review fixes

VladimirMakeev committed:
Add player buildings to Lua api

VladimirMakeev committed:
Fix edit box input

VladimirMakeev committed:
Add examples to documentation

VladimirMakeev committed:
Define noble action categories

VladimirMakeev committed:
Read custom noble action categories

VladimirMakeev committed:
Properly define GlobalVariables and support custom values

VladimirMakeev committed:
Define CEventEffectBase

VladimirMakeev committed:
Implement CEffectStealMarket

VladimirMakeev committed:
Implement CEffectGiveResources

VladimirMakeev committed:
Define INobleActionResult and its vftable

VladimirMakeev committed:
Implement noble action result 'steal from market'

VladimirMakeev committed:
Implement custom noble action 'steal from resource market'

VladimirMakeev committed:
Add missing include

VladimirMakeev committed:
Update documentation

VladimirMakeev committed:
Minor fixes
